From bd8f68140bc3e026ee27ea8e0898cedc7fd2a1fc Mon Sep 17 00:00:00 2001
From: Stefan Berger <stefanb@linux.ibm.com>
Date: Wed, 18 May 2022 21:20:35 -0400
Subject: [PATCH 79/81] evm: Adjust capability check for container root user

Adjust the check for capabilities inside the container so that a container
user with CAP_SYS_ADMIN or CAP_MAC_ADMIN can also write to the evm
securityfs file.

Signed-off-by: Stefan Berger <stefanb@linux.ibm.com>
---
 security/integrity/evm/evm_secfs.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/security/integrity/evm/evm_secfs.c b/security/integrity/evm/evm_secfs.c
index 7bce3320bb33..76ea682cf1c4 100644
--- a/security/integrity/evm/evm_secfs.c
+++ b/security/integrity/evm/evm_secfs.c
@@ -68,7 +68,8 @@ static ssize_t evm_write_key(struct file *file, const char __user *buf,
 	if (!ns_is_active(ns))
 		return -EPERM;
 
-	if (!capable(CAP_SYS_ADMIN) ||
+	if ((!mac_admin_ns_capable(current_user_ns()) &&
+	     !capable(CAP_SYS_ADMIN)) ||
 	    (ns->evm_initialized & EVM_SETUP_COMPLETE))
 		return -EPERM;
 
-- 
2.36.1

